src/patchelf.cc

```diff
@@ -1274,26 +1312,109 @@ void ElfFile<ElfFileParamNames>::rewriteHeaders(Elf_Addr phdrAddress)
        remapped. */
     for (unsigned int i = 1; i < rdi(hdr()->e_shnum); ++i) {
         auto &shdr = shdrs.at(i);
-        if (rdi(shdr.sh_type) != SHT_SYMTAB && rdi(shdr.sh_type) != SHT_DYNSYM) continue;
-        debug("rewriting symbol table section %d\n", i);
+
+        if (rdi(shdr.sh_type) != SHT_SYMTAB && rdi(shdr.sh_type) != SHT_DYNSYM) {
+            continue;
+        }
+
+        debug("rewriting symbol table (section %d)\n", i);
+        
         for (size_t entry = 0; (entry + 1) * sizeof(Elf_Sym) <= rdi(shdr.sh_size); entry++) {
             auto sym = (Elf_Sym *)(fileContents->data() + rdi(shdr.sh_offset) + entry * sizeof(Elf_Sym));
             unsigned int shndx = rdi(sym->st_shndx);
-            if (shndx != SHN_UNDEF && shndx < SHN_LORESERVE) {
-                if (shndx >= sectionsByOldIndex.size()) {
-                    fprintf(stderr, "warning: entry %d in symbol table refers to a non-existent section, skipping\n", shndx);
-                    continue;
+
+            if (shndx == SHN_UNDEF || shndx >= SHN_LORESERVE) {
+                continue;
+            }
+
+            if (shndx >= sectionsByOldIndex.size()) {
+                warn(
+                    "entry %d refers to a non-existent section (%d), skipping\n",
+                    entry,
+                    shndx
+                );
+
+                continue;
+            }
+
+            const std::string & section = sectionsByOldIndex.at(shndx);
+            assert(!section.empty());
+
+            auto newIndex = getSectionIndex(section);
+            auto newValue = rdi(sym->st_value);
+
+            if (ELF32_ST_TYPE(rdi(sym->st_info)) == STT_SECTION) {
+                newValue = rdi(shdrs.at(newIndex).sh_addr);
+            } else {
+                auto rs = replacedSections.find(section);
+
+                if (rs == replacedSections.end()) {
+                    // Entry refers to a section that didn't got moved around
+                    // (we still might have to update shndx, though)
+                } else {
+                    if (rdi(sym->st_value) < rs->second.prev_addr) {
+                        warn("entry %d refers to an address outside the section, skipping\n", entry);
+                        continue;
+                    } else {
+                        newValue = rdi(sym->st_value) - rs->second.prev_addr + rdi(shdrs.at(newIndex).sh_addr);
```

Comment:
This is part of the extra fix, for https://github.com/NixOS/patchelf/issues/75 - it doesn't seem to be required for the pcloud-nodejs problem, but I coded it before I stumbled upon the PHT thingie, that's why it's here.

Could be nice to keep, can be refactored into a separate MR as well, if review proves difficult here.